cli/src/commands/git.rs

```diff
@@ -1018,6 +1043,104 @@ fn cmd_git_submodule_print_gitmodules(
     Ok(())
 }
 
+fn cmd_git_submodule_clone(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    args: &GitSubmoduleCloneArgs,
+) -> Result<(), CommandError> {
+    let mut workspace_command = command.workspace_helper(ui)?;
+    let commit = workspace_command.resolve_single_rev(&args.revisions, ui)?;
+    let submodules = parse_commit_gitmodules(&workspace_command, &commit)?;
+    let submodule = match submodules.get(&args.name) {
+        Some(submodule) => submodule,
+        None => return Err(user_error(format!("{} is not a submodule", args.name))),
+    };
+
+    let mut tx = workspace_command.start_transaction(&format!("clone submodule {}", args.name));
+    let submodule_path = tx
+        .mut_repo()
+        .submodule_store()
+        .get_submodule_path(&args.name);
+    let submodule_path_existed = submodule_path.exists();
+    if submodule_path_existed {
+        if !is_empty_dir(&submodule_path) {
+            return Err(user_error(format!(
+                "Could not clone submodule: {} exists",
+                submodule_path.display()
+            )));
+        }
+    } else {
+        fs::create_dir(&submodule_path).unwrap();
+    }
+
+    let clone_result = do_submodule_clone(ui, command, &submodule_path, &submodule.url);
+    if clone_result.is_err() {
+        // Canonicalize because fs::remove_dir_all() doesn't seem to like e.g.
+        // `/some/path/.`
+        let canonical_submodule_path = submodule_path.canonicalize().unwrap();
+        if let Err(err) = fs::remove_dir_all(canonical_submodule_path.join(".jj")).and_then(|_| {
+            if !submodule_path_existed {
+                fs::remove_dir(&canonical_submodule_path)
+            } else {
+                Ok(())
+            }
+        }) {
+            writeln!(
+                ui,
+                "Failed to clean up {}: {}",
+                canonical_submodule_path.display(),
+                err
+            )
+            .ok();
+        }
+    }
+    writeln!(ui, "Cloned submodule \"{}\"", submodule.name)?;
+    // Unlike cmd_git_clone(), don't update the submodule's working copy.
+    // TODO do we need this transaction? We aren't changing branches
+    tx.finish(ui)?;
+    Ok(())
+}
+
+// This is mostly copied from do_git_clone(). We might want to dedup them in the
+// future.
+fn do_submodule_clone(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    submodule_path: &Path,
+    submodule_url: &str,
+) -> Result<(), CommandError> {
+    // TODO The submodule shouldn't have a workspace, but we don't have a good
+    // way to init a repo without one.
```

Comment:
Because `ReadonlyRepo::init()` has too many arguments?  I agree that it's a bit verbose, but you should be able to do something like this:
```
            ReadonlyRepo::init(
                &settings,
                &repo_dir,
                |store_path| -> Result<Box<dyn Backend>, BackendInitError> {
                    Ok(Box::new(GitBackend::init_internal(store_path)?))
                },
                ReadonlyRepo::default_op_store_factory(),
                ReadonlyRepo::default_op_heads_store_factory(),
                ReadonlyRepo::default_index_store_factory(),
                ReadonlyRepo::default_submodule_store_factory(),
            )
            .unwrap()
```